Monika After Story/game/zz_selector.rpy

```diff
@@ -3243,7 +3247,7 @@ screen mas_selector_sidebar(items, mailbox, confirm, cancel, restore, remover=No
                             xalign 0.5
 
                     for selectable in flt_items:
-                        if (item_type is None) or (selectable.selectable in filter_map[item_type]):
+                        if (mailbox.item_type is None) or (selectable.selectable in filter_map[mailbox.item_type]):
```

Comment:
The `in` check here is expensive since its going thru a list.

Rather than this, if a filter is set, we should set a screen var with the selectables from the `filter_map`, then have the search code (`selector_search_callback`) adjusted so it checks that screen var first. 

However, this would also need to manually trigger search to re-run whenever the filter is selected since the search callback is only on input change. I'm not sure if we save the search_query anywhere (@Booplicate might know) but that would have to be saved somewhere (ideally mailbox) so we manually trigger search.

Reply:
I want to clarify, you suggest to append the selected filter option to the search query?

Reply:
The search callback indeed can access the screen scope, so we can set/get a screen var (or mailbox var). Although, there's a few things I want to mention.

If the user searched, we shouldn't change the query. That's counterintuitive, plus we might not be able to remove that from the query later (say if the user removes/changes the filter).

While the search doesn't seem to cause any slowdown, it is kinda heavy. It checks name, id, type, and ex-props of **every provided item** (consider also all the callbacks it has to execute). Its flexibility comes at the cost of speed, I don't think we should rerun it on filter change. As of right now filtering seems to be more specific, but faster.

Reply:
ok going to have some definitions here:
* button filter - the filter buttons (aka the buttons that map to a list of selectables)
* search filter - the filtering that occurs when a search is run

>If the user searched, we shouldn't change the query. That's counterintuitive, plus we might not be able to remove that from the query later

I did not mean change the search query, I mean save the search query to retrigger the search on the button-filtered set of items. 

Right now the way this works is the search code filters down `items`, and then the loop that adds the selectables to the screen does another loop per item to check if its in the list of selectables from the button filter.

A better solution would be that the search (which is the most expensive) runs on the limited set of selectables given by the button-filter (which is a direct mapping) if possible. But the problem is that if you ran a search prevoiusly, then use the button filter, the search code isn't going to re-run since its a callback on the input field.

pasting from @ohnoitsavon 's DM, (btw just respond on here for documentation purposes):
>i have worked out exactly what you meant now in terms of "screen variable" and can see how it would work, however it is not possible as the search sub functions all revolve arround "MASSelectableImageButtonDisplayable" and the items are not that, they are just the normal selectables

Yup, not possible at least how I was originally thinking about it. But another option could be:  
* in the pre-screen label (`mas_selector_sidebar`), loop over `filter_map` and replace selectables with their disp equivalent in items - (sort of expensive, but its a run-once so cheaper in the long run). This will be what is passed into the actual screen.
* still have a screen var for the subset of selectable displayables - search code should check for this var first
* when the button filter is selected do 2 things:
    * set the screen var to the list of selectable disps based on `filter_map`
    * loop over `flt_items` and remove ones that don't belong - I think the loop is probably faster than re-running the search, but the  key point is that its only ran when a button filter is selected instead of constantly via screen refresh.

Reply:
@Booplicate @ThePotatoGuy I think I can do something very close to that, I am looking into converting them in mas_selector_sidebar.

Reply:
> A better solution would be that the search (which is the most expensive) runs on the limited set of selectables given by the button-filter (which is a direct mapping) if possible. But the problem is that if you ran a search prevoiusly, then use the button filter, the search code isn't going to re-run since its a callback on the input field.

Ah, okay that makes perfect sense now. Yeah, just save the query in the mailbox and call the search callback again as a button action.

> I am looking into converting them in mas_selector_sidebar.

Making displayables in the screen isn't ideal since that code will be: 1. rerun on every interaction, 2. rerun as we loop.

Reply:
updated my comment - I meant `mas_selector_sidebar_select`, which is the pre-loop label

Reply:
New commit submitted